### PR TITLE
describe texel formats (for storage texture types)

### PR DESCRIPTION
**Pull request by dneto0**
I'd like to describe textures in more detail so we can more precisely describe the return type on texture builtin functions, and the sampled component type on sampled texture types.
I'd like to point back to a table in the "Texel formats" section introduced in this PR.

### CODE REVIEW

File: wgsl/index.bs
Comment (mehmetoguzderin) [inline wgsl/index.bs:632]: Seems like this lacks 2D annotation right after `type(texel_format)` (all others do have)

File: wgsl/index.bs
Comment (mehmetoguzderin) [inline wgsl/index.bs:635]: Seems like this lacks 2D annotation right after `type(texel_format)` (all others do have)

**Review by mehmetoguzderin — CHANGES_REQUESTED** [review]

File: wgsl/index.bs
Comment (dj2) [inline wgsl/index.bs:553]: Is this a note, or a partial sentence?

File: wgsl/index.bs
Comment (dj2) [inline wgsl/index.bs:575]: I'd suggest changing this to 1 line per `td` to make it easier to read/work with

File: wgsl/index.bs
Comment (dj2) [inline wgsl/index.bs:584]: `ui` should be unsigned integer value no?

File: wgsl/index.bs
Comment (dj2) [inline wgsl/index.bs:585]: 32 bit signed integer value?

File: wgsl/index.bs
Comment (Kangz) [inline wgsl/index.bs:526]: nit: multi-dimensional grid (it can be 1, 2, 3 or even 4 if when we support 2D array multisampled textures)

File: wgsl/index.bs
Comment (Kangz) [inline wgsl/index.bs:531]: nit: isn't the arrayness just an extra dimension for the multi-dimensional grid?

File: wgsl/index.bs
Comment (Kangz) [inline wgsl/index.bs:536]: What about mip levels?

File: wgsl/index.bs
Comment (Kangz) [inline wgsl/index.bs:582]: I don't think that the WGSL spec should talk about the exact bit representation of texture, because there's weird formats like rgb9e5 and compressed ones. Only the component type matters, except for storage textures (which can have an exact bit representation described in this spec).

File: wgsl/index.bs
Comment (kvark) [inline wgsl/index.bs:527]: it's a bit weird to call them "special operations useful for rendering" and then follow with "those operations", which assumes the users know what they are. Would be good to provide examples here

File: wgsl/index.bs
Comment (kvark) [reply wgsl/index.bs:582]: I think there is still value in describing the formats we *do* know and guarantee, such as `8unorm`. For `rgb9e5` we can figure something out and say it's more platform-dependent.

Comment (Kangz) [reply wgsl/index.bs:582]: I'm not sure I understand, should the WGSL spec just tell you that when you sample a texture with floating point component type you get an `f32` and what `f32` you get is described by the texture sampling part of the WebGPU specification? This way WGSL wouldn't care if it is `r8unorm` `rg16float` or `bc1-rgba-unorm`?

Comment (kvark) [reply wgsl/index.bs:582]: well, I'm saying that it's generally useful that, for example, an r8unorm is one of the values between 0/255 and 255/255. This allows authors to adjust their expectations on sampling, or when writing to the color outputs.
Where exactly this is specified (WGSL or WebGPU spec) is not something that I feel strongly about.

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs]: partial sentence.  This was rewritten.

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs]: The only reason WGSL needs texel-formats is to parameterize storage textures. The sampled and depth textures do not, and I'd like to eliminate those weirdly-shaped texel formats from the WGSL spec entirely.

See the modifications to the texture-ro and texture-wo sections in this MR.

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs]: I wanted to defer those to the details of builtin functions.  That's why later in this introductory section we get into the fact there's no direct access to texture storage and that everything is determined by the set of available texture builtins.

Maybe I can think of something that points forward to the builtins without going into details.

Frankly, I'm concerned about having to describe sampling operations. I'd rather not do that in the WGSL spec.  Like, there's a *lot* of content in Vulkan's "14. Image operations". https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#textures

**Comment by dneto0** [comment]
HTML preview is at https://gistpreview.github.io/?4c60af08a8b1cca4501abecc93beae3f

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs]: I was hoping to treat the array index specially. The regular grid coordinates have interpolation applied to them, but the array index does not.  I suggest we add a note to revisit this editorially

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs]: Indeed.  I've yet to capture that.

**Comment by dneto0** [comment]
TODO:  capture the concept of mip-levels.

File: wgsl/index.bs
Comment (kvark) [inline wgsl/index.bs:615]: We can't have that in the language. Is it truly undefined? Could you make an issue/mini-investigation about this?

File: wgsl/index.bs
Comment (kvark) [inline wgsl/index.bs]: it seems to me that we'd need to remove any mention of this SPIR-V capability, because it doesn't translate to what WebGPU can support, see https://gpuweb.github.io/gpuweb/#texture-format-caps

File: wgsl/index.bs
Comment (kvark) [inline wgsl/index.bs:830]: it says "conversion function" and then "transfer function". Is this the same thing?

Comment (kvark) [inline wgsl/index.bs:830]: it's not clear how the inversion works, i.e. would the values be rounded or truncated, for example

**Review by kvark — CHANGES_REQUESTED** [review]

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs]: I did that for the headings.
But I left the rows as-is.  I find it easier to both edit and scan for differences when they're horizontally organized

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs:615]: I filed #1115 for this.

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs]: This table is for translation _to_ SPIR-V.  Some of the image formats that I think are to be supported by WebGPU  do require the extended image formats capability.

File: wgsl/index.bs
Comment (kvark) [reply wgsl/index.bs:615]: Let's have an `Issue(gpuweb/1115): xxx` for it here (or whatever the syntax is)

File: wgsl/index.bs
Comment (kvark) [reply wgsl/index.bs]: I don't think any of our formats require this capability today.

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs]: The bullet points below describe dimensionality, and say it's 1,2,3 dimensions. There is also array-size, when the texture is arrayed.

Note that the compute shader workgroup-grid is 3-dimensional without explicitly calling it out as multi-dimensional.

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs]: After removing the texel formats that require SPIR-V capability StorageImageExtendedFormats, we're left with:

* r32uint
* r32sint
* r32float
* rgba8unorm
* rgba8snorm
* rgba8uint
* rgba8sint

I can delete all the other ones from the tables, and make this much simpler.

Comment (kvark) [reply wgsl/index.bs]: where did 32-bit RG and RGBA formats go?

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs]: After that discussion evolved, I've made missing green and blue map to 0, and alpha to 1.

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs]: Ah, ok.  I've updated to eliminate all those extended formats.

**Comment by dneto0** [comment]
The preview has been updated: https://gistpreview.github.io/?4c60af08a8b1cca4501abecc93beae3f#texel-formats

**Comment by dneto0** [comment]
Rebased against main, and I've updated the gist preview

**Comment by kvark** [comment]
My comment may have been lost, so repeating it here. Why did you remove RG32 and RGBA32 formats?

File: wgsl/index.bs
Comment (dj2) [reply wgsl/index.bs]: Looks like `rgba32float` and `rgba32ui`, `rgba32i` are missing from list above but all of the `rg` formats appear to be in `StorageImageExtendedFormats`.  https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Image_Format

**Comment by dj2** [comment]
The `RG32` formats require `StorageImageExtendedFormats`. Looks like the `RGBA32` ones are missing.

https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Image_Format

File: wgsl/index.bs
Comment (kvark) [reply wgsl/index.bs]: Interesting! So this is different from what Vulkan requires. I guess the `StorageImageExtendedFormats` on the SPIR-V side becomes an implementation detail, and we don't need to care about it on the API side.
Please just follow the table we have in the spec - https://gpuweb.github.io/gpuweb/#plain-color-formats

Comment (dneto0) [reply wgsl/index.bs]: I had been reading the required format support for Vulkan. But I have not been following what was worked out already for WebGPU. 
I'll follow the WebGPU table. Thanks!

**Comment by kvark** [comment]
@dj2 see https://github.com/gpuweb/gpuweb/pull/1112#discussion_r498320940

**Comment by dneto0** [comment]
I've made the list of storage texture formats match what's already listed in the WebGPU table.

That adds the Rg32* formats. Even though those have mandatory in Vulkan, they still require the StorageImageExtendedFormats capability in SPIR-V.  So I added back the "SPIR-V enabling capability" column. 

I've updated the HTML preview: https://gistpreview.github.io/?4c60af08a8b1cca4501abecc93beae3f#texel-formats 

I've added TODOs to move the description about how storage texture accesses work down to the texture builtins section.  The write access needs a more explicitly spelled out description.

**Review by kvark — APPROVED** [review]

**Comment by dneto0** [comment]
I've rebased against main, and collapsed the previous contents into a single commit.

**Comment by dneto0** [comment]
I've updated the HTML preview at https://gistpreview.github.io/?4c60af08a8b1cca4501abecc93beae3f 

Note that other texture type changes are already in flight: #1180, #1182  

I'd like to get these changes approved, modulo those other changes, which are basically orthogonal.

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs]: I added mip levels now.  Thanks!

**Comment by dneto0** [comment]
I've rebased on top of recent renamings of storage texture types, and:

- added cross-references to WebGPU concepts
- added mip levels to the list of texture properties.
- fixed wordings.

Please take a look.

**Comment by grorg** [comment]
Discussed at [2020-11-17 meeting](https://docs.google.com/document/d/1JMqCbDXyuEvjH8GalT8_bEJ9vfejMo4GMIsAVaXnKvI/).

File: wgsl/index.bs
Comment (kvark) [inline wgsl/index.bs:722]: Storage class name needs an update.
In WebGPU spec we'd reference those as something like {{GPUStorageClass/"handle'"}}, which allows `bikeshed` to generate a reference link

**Review by kvark — APPROVED** [review]

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs:722]: I'll fix this to 'handle' now.

The  {{Storage classes/handle}} construct requires a bigger change to the storage class table.

It looks like the syntax that will work here is    [=storage classes/handle=].
I'd like to do the whole storage classes table at once, but that will generate a conflict with this PR.  So I'll do that in a followup.

Comment (dneto0) [reply wgsl/index.bs]: I filed #1232 for followup

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs:938]: Fixed in dc7f7538a

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs:941]: Fixed in dc7f7538a

**Review by mehmetoguzderin — APPROVED** [review]
Looks great!

**Pull request merged**
